### ISSUE TITLE
Change Request: Ajv makeValidate is blocked in WebExtension context

### ISSUE THREAD

**Issue by erosman** [eslint/eslint#19635]
### ESLint version

9.24.0

### What problem do you want to solve?

An issue occurs in bundling ESLint for WebExtension use.

It appears that ESLint is depending [Ajv](https://github.com/eslint/eslint/blob/910bd13c4cb49001f2a9f172229360771b857585/package.json#L120) to [validate ESLint configurations](https://github.com/eslint/eslint/blob/910bd13c4cb49001f2a9f172229360771b857585/lib/config/rule-validator.js#L158) by matching them to a JSON schema. Ajv [uses Function()](https://github.com/ajv-validator/ajv/blob/82735a15826a30cc51e97a1bbfb59b3d388e4b98/lib/compile/index.ts#L171) to compile JSON schemas into JavaScript functions.

The `new Function()` is blocked by the browser CSP which results in a `throw`.

See also: [Errors with eslint-linter-browserify](https://github.com/UziTech/eslint-linter-browserify/issues/519)

### What do you think is the correct solution?

An alternative method (for the WebExtension context) would be greatly beneficial.

### Participation

- [ ] I am willing to submit a pull request for this change.

**Comment by fasttime** [comment]
Hi @erosman! It's not clear to me how to reproduce the problem you are experiencing. Keep in mind that there is no guarantee that ESLint will run in environments with deliberate restrictions such as browser extensions. I can see that if the `Function` constructor is suppressed in Node.js with the `--disallow-code-generation-from-strings` flag, `eslint` runs into an error:

```console
$ node --disallow-code-generation-from-strings bin/eslint.js  
(node:75435) ESLintIgnoreWarning: The ".eslintignore" file is no longer supported. Switch to using the "ignores" property in "eslint.config.js": https://eslint.org/docs/latest/use/configure/migration-guide#ignoring-files
(Use `node --trace-warnings ...` to show where the warning was created)
Error compiling schema, function code:  var validate = function(data, dataPath, parentData, parentDataProperty, rootData) { 'use strict';  var vErrors = null;  var errors = 0;      if (rootData === undefined) rootData = data;  if (Array.isArray(data)) {   if (  data.length > 0) {  validate.errors = [ { keyword: 'maxItems' , dataPath: (dataPath || '') + "" , schemaPath: '#/maxItems' , params: { limit: 0 }  , message: 'should NOT have more than 0 items'  , schema:  0         , parentSchema: validate.schema , data: data  } ]; return false; }  else {    if (  data.length < 0) {  validate.errors = [ { keyword: 'minItems' , dataPath: (dataPath || '') + "" , schemaPath: '#/minItems' , params: { limit: 0 }  , message: 'should NOT have fewer than 0 items'  , schema:  0         , parentSchema: validate.schema , data: data  } ]; return false; }  else {   }}  }  else {  validate.errors = [ { keyword: 'type' , dataPath: (dataPath || '') + "" , schemaPath: '#/type' , params: { type: 'array' }  , message: 'should be array'  , schema: validate.schema.type , parentSchema: validate.schema , data: data  } ]; return false;  }  if (errors === 0) {  }  validate.errors = vErrors;  return errors === 0;        }; return validate;

Oops! Something went wrong! :(

ESLint: 9.25.1

Error: Error while processing options validation schema of rule 'json/no-duplicate-keys': Code generation from strings disallowed for this context
    at RuleValidator.validate (.../eslint/lib/config/rule-validator.js:161:12)
    at new Config (.../eslint/lib/config/config.js:255:18)
    at [finalizeConfig] (.../eslint/lib/config/flat-config-array.js:212:10)
    at FlatConfigArray.getConfigWithStatus (.../eslint/node_modules/@eslint/config-array/dist/cjs/index.cjs:1263:55)
    at FlatConfigArray.getConfig (.../eslint/node_modules/@eslint/config-array/dist/cjs/index.cjs:1281:15)
    at entryFilter (.../eslint/lib/eslint/eslint-helpers.js:288:27)
    at async NodeHfs.<anonymous> (file://.../eslint/node_modules/@humanfs/core/src/hfs.js:574:24)
    at async NodeHfs.walk (file://.../eslint/node_modules/@humanfs/core/src/hfs.js:614:3)
    at async globSearch (.../eslint/lib/eslint/eslint-helpers.js:329:20)
    at async Promise.allSettled (index 0)
```

I imagine this is related to your problem, but we will need instructions on how to reproduce the error to look into this.

**Comment by erosman** [comment]
PS. As asked in https://github.com/eslint/eslint/discussions/19641, does a standalone [Linter](https://eslint.org/docs/latest/integrate/nodejs-api#linter) need Ajv `makeValidate`?

**Comment by fasttime** [comment]
Thanks for the info! It seems that Ajv relies on functions created dynamically with the `Function` constructor to validate JSON schemas, and if the `Function` constructor is set to throw an error, for example per CSP, then the schema validation fails. In ESLint, Ajv is used to validate rule schemas. In legacy config mode, this validation is only done in the `ESLint`/`LegacyESLint` constructor, whereas in the new config mode, rules are validated when the configuration is first applied, which could be in the `Linter` as well. This explains why the error only occurs in ESLint v9.

**Comment by fasttime** [comment]
> An alternative method (for the WebExtension context) would be greatly beneficial.

Do you know an alternative method that preserves the validation?

**Comment by erosman** [comment]
I am not an expert at all but I have been looking at other validators. Others don’t "appear" to use `new Fuicntio()` (but I am not sure). 🤔 

Ajv
https://github.com/ajv-validator/ajv/blob/82735a15826a30cc51e97a1bbfb59b3d388e4b98/lib/compile/index.ts#L171-L172

Joi 
https://github.com/hapijs/joi/blob/b32b80d87e30e83596e9e5f427dd361e3bf7602c/lib/types/any.js#L44

yup
https://github.com/jquense/yup/blob/ef5303025c38e6e0dc0de53c990e7277cc74164e/src/schema.ts#L558

Zod
I didnt find exactly where it parses schema

jsonschema
https://github.com/tdegrunt/jsonschema/blob/f65c769eac5eda621f6d09f4df86c8fc268d6344/lib/index.d.ts#L56

**Comment by fasttime** [comment]
> I am not an expert at all but I have been looking at other validators. Others don’t "appear" to use `new Fuicntio()` (but I am not sure). 🤔

There have been several attempts to replace Ajv, but in the end, they were rejected for not being backwards-compatible:

* #10153
* #13911
* #19340

At this time, the only solution I can think of is skipping schema validation if `Function("")` throws. Even so, running ESLint in a browser extension is not something we currently support, so there could be as well other issues.

**Comment by erosman** [comment]
> There have been several attempts to replace Ajv, but in the end, they were rejected for not being backwards-compatible:

Again, I am not an expert, but isn’t the validator’s job returning `true/false`?

```js
const validate = ajv.compile(schema)
const valid = validate(data)
if (!valid) console.log(validate.errors)
```

Shouldn’t replacing the validator be inconsequential?

```js
const validate = NotAjv.compile(schema)
const valid = validate(data)
if (!valid) console.log(validate.errors)
```

> the only solution I can think of is skipping schema validation if `Function("")` throws.

Would [Linter](https://eslint.org/docs/latest/integrate/nodejs-api#linter) work without it?
Can I replace it (e.g. by making it return true) from my side?
Can it be changed to behave like ESLint v8?
Can the call to it be commented out?
_(I am not familiar with ESLint's code.)_

> Even so, running ESLint in a browser extension is not something we currently support

I understand, but bearing the environment in mind for future developments, should be productive.
As an idea, I would suggest developing [Linter](https://eslint.org/docs/latest/integrate/nodejs-api#linter) into a more independent module.

I use ESLint 9 + VS Code for my own programming work and have been very happy with ESLint.
I am in the process of implementing ESLint v8, in conjunction with Monaco Editor in one of my browser extensions (replacing the current CodeMirror 5 + JSHint) and so far, it is working fine.
I would prefer to use the latest ESLint though.

**Comment by fasttime** [comment]
> Again, I am not an expert, but isn’t the validator’s job returning `true/false`?
> 
> const validate = ajv.compile(schema)
> const valid = validate(data)
> if (!valid) console.log(validate.errors)

The validator creates an error message to report what is wrong in a rule's config, for example "Value should NOT have more than 2 items."

> Shouldn’t replacing the validator be inconsequential?
> 
> const validate = NotAjv.compile(schema)
> const valid = validate(data)
> if (!valid) console.log(validate.errors)

It should. In fact there seems to be no replacement for Ajv 6 that keeps its functionality unchanged.

> > the only solution I can think of is skipping schema validation if `Function("")` throws.
> 
> Would [Linter](https://eslint.org/docs/latest/integrate/nodejs-api#linter) work without it? Can I replace it (e.g. by making it return true) from my side? Can it be changed to behave like ESLint v8? Can the call to it be commented out? _(I am not familiar with ESLint's code.)_

You can certainly create your own fork of ESLint and make any changes you want as long the MIT license is kept unchanged. See https://github.com/eslint/eslint/blob/main/LICENSE. Ajv is invoked here:

https://github.com/eslint/eslint/blob/129882d2fdb4e7f597ed78eeadd86377f3d6b078/lib/config/rule-validator.js#L158

> I understand, but bearing the environment in mind for future developments, should be productive. As an idea, I would suggest developing [Linter](https://eslint.org/docs/latest/integrate/nodejs-api#linter) into a more independent module.
> 
> I use ESLint 9 + VS Code for my own programming work and have been very happy with ESLint. I am in the process of implementing ESLint v8, in conjunction with Monaco Editor in one of my browser extensions (replacing the current CodeMirror 5 + JSHint) and so far, it is working fine. I would prefer to use the latest ESLint though.

There are plans to refactor the `Linter` class into a new API that could be used both in the browser and on a file system. You may want to read the discussion in #16557 (it's read-only now).

If you need advice about solving some specific problem it's better to open a [discussion](https://github.com/eslint/eslint/discussions) or use [the Discord chat](https://eslint.org/chat) because lengthy comments in a change request make it difficult for other users to understand what is being requested.

**Comment by erosman** [comment]
> You can certainly create your own fork of ESLint and make any changes you want as long the MIT license is kept unchanged. See https://github.com/eslint/eslint/blob/main/LICENSE. Ajv is invoked here:

I was actually asking for guidance. I wasn’t planning on forking ESLint. 
What I had in mind was to comment out or edit slightly to bypass the `new Fucntion()` during the bundling process.

> If you need advice about solving some specific problem it's better to open a [discussion](https://github.com/eslint/eslint/discussions) or use [the Discord chat](https://eslint.org/chat) because lengthy comments in a change request make it difficult for other users to understand what is being requested.

You are right. 
I had already done that though (#19641).
